### ISSUE TITLE
libgap: add (optional?) limb size argument to GAP_MakeObjInt

### ISSUE THREAD

**Issue by embray** [gap-system/gap#4209]
### Observed behaviour

`GAP_MakeObjInt` assumes limbs are `UInt`-sized (which is enforced at compile time to be the same size as `mp_limb_t`).  However, in Python limbs of bigints are either 32-bit or 16-bit integers depending on how Python was compiled, so converting a Python int to a GAP int requires some non-trivial conversion.

### Expected behaviour

It would be nice if `GAP_MakeObjInt` accepted a limb size argument, as well as a *nails* argument (significant bits to ignore, typically 2 or 1 on Python).  It could then use these arguments to pass through `mpz_import` and then convert the resulting `mpz_t` to a GAP integer.  As it is, this needs to be done manually.

**Comment by fingolfin** [comment]
I don't like a new argument to `GAP_MakeObjInt`, as this is already used in several downstream projects. But of course a new API is added.

Some free thought flow musings follow:

However, I wonder how much use this has beyond Python? Almost everything else seems to be using GMP, so can call `GAP_MakeObjInt` directly.

Honest question: could you perhaps call `mpz_import` in your code and then pass the result to `GAP_MakeObjInt`?

On the other hand, perhaps such an API would be useful for GAP itself, internally: one thing that's bothered me for some time about the GAP compiler is the following: given a program that contains a larger integer literal, it produces different C code depending on whether it was compiled in 32bit or 64bit mode. And that C code in turn will contain `#ifdef` to wrap different code for 32/64 bit mode. This would be much better if we refactored the code to use (wrappers around) `mpz_export` and `mpz_import`, to serialize the data into (32 bit) int arrays (plus some extra work for immediate integers, whose range also differs between 32/64 bit ways, but that's details that can be dealt with in various ways).

Gotta think a little bit about it.

**Comment by fingolfin** [comment]
Perhaps @ChrisJefferson has some thought as well

**Comment by ChrisJefferson** [comment]
I'd be happy to see a new function. I'm not 100% clear on the various arguments to GMP -- I'd prefer in we didn't end up with an interface that was very hard-wired to "Do what GMP does exactly" (unless that's also easy to document and implement for something which is not GMP), because there might be a time when we don't want to use GMP (for example, in the past when I've compiled GAP to Javascript I used GAP's older built-in large ints, as I couldn't get GMP to compile to Javascript).

**Comment by embray** [comment]
> Honest question: could you perhaps call mpz_import in your code and then pass the result to GAP_MakeObjInt?

Yes, that's what I'm doing now.  Since that's necessary anyways (either built into GAP, or done just for Python), I guess there's not much advantage to adding a new function to GAP or changing the existing one.  I agree with your point that as far as I know it's just Python that needs this.
